enterprise/frontend/src/embedding-sdk/store/auth.ts

```diff
@@ -0,0 +1,158 @@
+import type {
+  EmbeddingSessionToken,
+  FetchRequestTokenFn,
+  SDKConfig,
+} from "embedding-sdk";
+import { getIsLocalhost } from "embedding-sdk/lib/is-localhost";
+import type { SdkStoreState } from "embedding-sdk/store/types";
+import api from "metabase/lib/api";
+import { createAsyncThunk } from "metabase/lib/redux";
+import { refreshSiteSettings } from "metabase/redux/settings";
+import { refreshCurrentUser } from "metabase/redux/user";
+
+import { getOrRefreshSession } from "./reducer";
+import { getFetchRefreshTokenFn } from "./selectors";
+
+export const initAuth = createAsyncThunk(
+  "sdk/token/INIT_AUTH",
+  async (sdkConfig: SDKConfig, { dispatch }) => {
+    // Setup JWT or API key
+    const isValidJwtConfig =
+      sdkConfig.jwtProviderUri && sdkConfig.jwtProviderUri?.length > 0;
+    const isValidApiKeyConfig = sdkConfig.apiKey && getIsLocalhost();
+
+    if (isValidJwtConfig) {
+      // JWT setup
+      api.onBeforeRequest = async () => {
+        const session = await dispatch(
+          getOrRefreshSession(sdkConfig.jwtProviderUri!),
+        ).unwrap();
+        if (session?.id) {
+          api.sessionToken = session.id;
+        }
+      };
+      // verify that the session is actually valid before proceeding
+      await dispatch(getOrRefreshSession(sdkConfig.jwtProviderUri!)).unwrap();
+    } else if (isValidApiKeyConfig) {
+      // API key setup
+      api.apiKey = sdkConfig.apiKey;
+    }
+    // Fetch user and site settings
+    const [user, siteSettings] = await Promise.all([
+      dispatch(refreshCurrentUser()),
+      dispatch(refreshSiteSettings({})),
+    ]);
+
+    if (!user.payload) {
+      // The refresh user thunk just returns null if it fails to fetch the user, it doesn't throw
+      const error = new Error(
+        "Failed to fetch the user, is the session valid?",
+      );
+      console.error(error);
+      throw error;
+    }
+    if (!siteSettings.payload) {
+      const error = new Error(
+        "Failed to fetch the site settings, is the session valid?",
+      );
+      console.error(error);
+      throw error;
+    }
+  },
+);
+
+export const refreshTokenAsync = createAsyncThunk(
+  "sdk/token/REFRESH_TOKEN",
+  async (url: string, { getState }): Promise<EmbeddingSessionToken | null> => {
+    // The SDK user can provide a custom function to refresh the token.
+    const customGetRefreshToken = getFetchRefreshTokenFn(
+      getState() as SdkStoreState,
+    );
+
+    const getRefreshToken = customGetRefreshToken ?? defaultGetRefreshTokenFn;
```

Comment:
It was like that before, I only moved the logic: https://github.com/metabase/metabase/blob/fedd1466998592b8733d7d891af6b6ef262599b9/enterprise/frontend/src/embedding-sdk/store/reducer.ts#L62-L64
I agree it may be more logic for the built in function to just be a default, but right now i'm using this behaviour to tell the user if the wrong values are coming from the custom fetch function or from the endpoint.

If that's ok i'll keep this logic as is for now, I plan on trying to figure out if we can avoid storing the function on redux at all as may make the debugger misbehave: https://redux.js.org/style-guide/#do-not-put-non-serializable-values-in-state-or-actions

Reply:
ahh. That makes sense about storing non serializable data in redux. Feels free to keep this as is.

Reply:
Yeah a context would be good, and I think it may be maybe easier than redux as we wouldn't even need to dispatch actions to sync the values, we could just forward what we receive from the props.
I'll see if it breaks anything when I'll have some time after the sdk fixes